### ISSUE TITLE
Unsoundness in union of object types

### ISSUE THREAD

**Issue by hanayashiki** [microsoft/TypeScript#45614]
# Bug Report

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `union types`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=17&ssc=2&pln=1&pc=1#code/C4TwDgpgBAalC8UDeAoK6oEMBcy0YIGNcBnYAJwEsA7AcwBp90BfRggI12oFcBbdiOUbMoAHzwcufAUJTMUKQgHtqZKADdccRKgI4o3agBMIAMxoQjbDJygBGOQsqmoACgDkmd1BoaAlBIYyqpKADYQAHShSrSu6hGYEYR+ANxyQA)

### 💻 Code

```ts
type V = {
    a: {
        c: string,
    },
    b: number,
} | {
    b: number,
}

const v: V = {
    a: undefined,
    b: 1
}

if ('a' in v) {
    console.log(v.a.c);  // In runtime there would be an error of accessing undefined
}
```

### 🙁 Actual behavior

No error is found

### 🙂 Expected behavior

I know TS is not built to be sound, and it is more a static analytic tool for JavaScript. But is this behavior expected?

In this case, `'a' in v` does not indicate `v.a` is `{ a: {.c: string } }`, because `v` could be anything containing `b: number`, but we don't make any assumptions about `a`, since as long as `b: number` exists then `v is V`.

I also find if we initialize `v` as 

```
const v: V = {
    a: 0,
    b: 1
}
```

It returns with an error `Type 'number' is not assignable to type '{ c: string; }'.`, even though, it is a subtype of `{ b: number }`.  

So, it means "only undefined can be assigned to `a`"? 

It seems `a: undefined` is a special case for object literals and union types, but it this documented anywhere?

Thanks in advance.

**Comment by MartinJohns** [comment]
Duplicate of #34975.
